controller/lib/hal/hal_stm32_regs.h

```diff
@@ -430,29 +447,95 @@ struct DMA_Regs {
       REG priority : 2; // priority level 0b00 - low, 0b11 - high
       REG mem2mem : 1;  // memory-to-memory mode
       REG rsvd : 17;
-    } config;  // channel x configuration register
-    REG count; // channel x number of data to transfer register
-    REG pAddr; // channel x peripheral address register
-    REG mAddr; // channel x memory address register
-    REG rsvd;  // reserved
+    } config;             // channel x configuration register
+    REG count;            // channel x number of data to transfer register
+    volatile void *pAddr; // channel x peripheral address register
+    volatile void *mAddr; // channel x memory address register
+    REG rsvd;             // reserved
   } channel[7];
   REG rsvd[5]; // reserved
-  struct {
-    REG c1s : 4;
-    REG c2s : 4;
-    REG c3s : 4;
-    REG c4s : 4;
-    REG c5s : 4;
-    REG c6s : 4;
-    REG c7s : 4;
-    REG rsvd : 4;
+  union {
+    struct {
+      REG c1s : 4;
+      REG c2s : 4;
+      REG c3s : 4;
+      REG c4s : 4;
+      REG c5s : 4;
+      REG c6s : 4;
+      REG c7s : 4;
+      REG rsvd : 4;
+    };
+    REG r;
   } chanSel; // channel selection register
 };
 inline DMA_Regs *const DMA1_BASE = reinterpret_cast<DMA_Regs *>(0x40020000);
 inline DMA_Regs *const DMA2_BASE = reinterpret_cast<DMA_Regs *>(0x40020400);
 
+// Select the source for a DMA channel
+// @param dma Address of DMA registers
+// @param chan DMA channel to modify.  Channels are numbered from 0
+// @param selection Selects which peripherial request to map to the channel
+inline void DMA_SelectChannel(DMA_Regs *const dma, DMA_Chan chan,
+                              int selection) {
+  selection &= 0x0F;
+
+  int x = 4 * static_cast<int>(chan);
+
+  uint32_t val = dma->chanSel.r;
+  val &= ~(0xF << x);
+  val |= selection << x;
+  dma->chanSel.r = val;
```

Comment:
I actually tried writing it as a switch statement over the channels, but ran into trouble with the very strict compile flags we're using.  When I tried to set the 4-bit value from an integer the compiler wouldn't let me because I could lose precision.  I wasn't sure how to cast the int to a 4-bit value, so ended up going this way.
I actually thought the case statement approach was easier to read so if you know how to get around that I'd happily switch back.

Reply:
I think the RHS has to match the type of REG, which is uint32.  If that doesn't work, can you show me the code and the error?